_posts/2024-01-11-gsoc2023_marinegor.md

```diff
@@ -0,0 +1,169 @@
+ - `_compute()`: run `_single_frame` on a list of frames, but without running `_conclude`
+ - `_get_aggregator()`: get an object to aggregate the run results with, making them compatible with subsequent `_conclude`
+ - class property `available_backends()`: get list of `str` values that describe available backends for a given subclass
+
+I've also added `ParallelExecutor` and `ResultsGroup` classes that abstract away parallel execution and results aggregation, respectively. And finally, I added `multiprocessing` and `dask`/`dask.distributed` backends that reportedly speed up the analysis!
+
+## The current state.
+Currently, changes to the `AnalysisBase` are almost finalized. One thing that holds it back is some CI/CD issues causing tests to timeout, but all `AnalysisBase`-related tests run both locally and on CI/CD system.
+
+## What's left to do.
+Optional thing suggested within my proposal was to actually add parallelization to the subclasses, and update tests accordingly. It turned out to be a tedious task, but finally the mechanism is finalized and described in `MDAnalysisTests/analysis/conftest.py`. It automatically generates each subclass fixtures for testing, and updating tests accordingly is fairly simple yet tedious. After updating all the tests, the library will be fully equipped with the parallel execution mechanisms for those classes that allow it.
+
+
+## What code got merged (or not) upstream.
+The main changes are summarized in the [main pull-request](https://github.com/MDAnalysis/mdanalysis/pull/4162) of the project. They mostly involve changes to `package/analysis/base.py`, as well as installation and CI/CD configuration files. Also, there are example changes in `package/analysis/rms.py` introducing parallelization into RMSD and RMSF subclasses, showcasing the changes to be made in order to add parallelization to a certain class.
+
+
+## How others can use it
+
+Let's imagine we've just learned that MDAnalysis now supports parallelization, and want to calculate RMSD of our large trajectory faster using multiple cores on our machine. Imagine we have a 16-core CPU Linux workstation with an SSD drive and a 1 us trajectory of lysozyme in water. Like this:
+
+
+```python
+import MDAnalysis as mda
+from MDAnalysis.analysis import rms
+
+prefix = "./large_data/"
+traj, top = f"{prefix}/md_0_1.xtc", f"{prefix}/md_0_1.gro"
+
+u = mda.Universe(top, traj)
+```
+
+First we want to get a reference structure by taking the average of all frames. Like this:
+
+```python
+from MDAnalysis.analysis.align import AverageStructure
+
+avg = AverageStructure(mobile=u).run(backend='multiprocessing', n_workers=16)
+```
+
+but we get this:
+
+```python
+ValueError                                Traceback (most recent call last)
+Cell In[11], line 2
+      1 from MDAnalysis.analysis.align import AverageStructure
+----> 2 avg = AverageStructure(mobile=u).run(backend='multiprocessing', n_workers=16)
```

Comment:
```suggestion
----> 2 avg = AverageStructure(mobile=u).run(backend='multiprocessing', n_workers=16)
```
For some reason, syntax highlighting doesn't seem to like this line, and messes up the highlighting of all subsequent code blocks/text. Maybe it's just sufficient to remove `---->`?

Reply:
I tried removing it - hopefully that works.